### PR TITLE
Fix remaining fee calculation

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#1561 

# Overview

- 938a3a034ca39ff2e821a109fb5ae8130667629a
  :round_pushpin: **add regression test triggering the unbalanced invariant**
    currently fails as expected with:

  ```
  Generated an unbalanced tx! Too much left for fees : fee (raw) = 164282 : fee (dangling) = 167722 , diff = 200000
  selection = inputs: - 1st 30 (~ 1000000 @ 66616b65...72657373)
  outputs: - 800000 @ 66616b65...72657373
  change: []

  CallStack (from HasCallStack):
    error, called at src/Cardano/Wallet/Primitive/Fee.hs:330:18 in cardano-wallet-core-2020.4.7-Be8Kf2A2Kpl7z2pVyjHPdZ:Cardano.Wallet.Primitive.Fee
    remainingFee, called at src/Cardano/Wallet/Primitive/Fee.hs:207:22 in cardano-wallet-core-2020.4.7-Be8Kf2A2Kpl7z2pVyjHPdZ:Cardano.Wallet.Primitive.Fee
  ```

- 1041fac37f0d87244d1a31dc90b8d1c62fa472e7
  :round_pushpin: **refactor: merge 'remainingFee' and 'rebalanceChangeOutput'**
  The former actually makes assumption based on the output of the later. Testing these
function in isolation makes little sense. Grouping them into one function makes it
easier to understand the surrounding context and what is happening.

Further refactoring is needed in order to now fix the issue with dangling fee

- 1758a306fbbeb234cbf48e43bcb7b9accc62aaca
  :round_pushpin: **use dustThreshold instead of 'Coin 0'**
  They should actually have the same semantic within this context.

- 7194126ca7933e08a9cd2528c94ef689623cb0f6
  :round_pushpin: **fix remaining fee calculation**
  
- a17462cf42a0e275d866d0bb7c2b001d5940ce42
  :round_pushpin: **fix unit test now correctly divying fee across change outputs**
  
- 15341246a634af562a4bf3984ecc147361f55e99
  :round_pushpin: **property test fee balancing**
  And allow transaction to be unbalanced if the node allows it

# Comments

### CODE REVIEW

**Comment by KtorZ** [comment]
bors try

File: lib/core/src/Cardano/Wallet/Primitive/Fee.hs [hunk 1]
```diff
@@ -253,13 +244,39 @@ coverRemainingFee (Fee fee) = go [] where
...
+        Fee φ_original = estimateFee opts s
+        -- The initial amount left for fee (i.e. inputs - outputs)
+        δ_original = inputBalance s - (outputBalance s + changeBalance s)
+
+        -- The new amount left after balancing (i.e. φ_original)
+        Fee φ_dangling = estimateFee opts sDangling
+        -- The new requested fee after adding the output.
+        δ_dangling = φ_original -- by construction of the change output
+
+        extraChng = Coin (δ_original - φ_original)
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/Fee.hs]: :+1:

**Review by paweljakubas — APPROVED** [review]
looks fine

File: lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs [hunk 2]
```diff
@@ -161,7 +167,7 @@ spec = do
             }) (Right $ FeeOutput
             { csInps = [20,20,20]
             , csOuts = [14,18,19]
-            , csChngs = [6]
+            , csChngs = [4,1,1]
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:170]: NOTE: To be fixed.

**Pull request merged**